core/data/pack/types.go

```diff
@@ -48,65 +50,79 @@ type types struct {
 // newTypes constructs a new empty type registry.
 func newTypes(forceDynamic bool) *types {
 	return &types{
-		entries:      []*ty{},
+		entries:      []*ty{nil},
 		byName:       map[string]*ty{},
 		forceDynamic: forceDynamic,
 	}
 }
 
-// addMessage adds a registry entry for a given message if needed.
-// It returns the registry entry, and a bool that is true if the entry
-// was newly added.
-func (t *types) addMessage(msg proto.Message) (ty, bool) {
-	return t.addType(reflect.TypeOf(msg).Elem())
+// addForMessage adds all types required to serialize the message.
+// The callback will be called for each newly added type.
+func (t *types) addForMessage(ctx context.Context, msg proto.Message, cb func(t *ty) error) (*ty, error) {
+	return t.addForType(ctx, reflect.TypeOf(msg), cb)
 }
 
-// addNameAndDesc adds a dynamic type by name and descriptor.
-// It uses the proto type registry to look up the name.
-func (t *types) addNameAndDesc(name string, desc *descriptor.DescriptorProto) (ty, bool) {
-	typ := proto.MessageType(name)
-	if typ == nil || t.forceDynamic {
-		if desc == nil {
-			return ty{}, false
-		}
-		create := func() proto.Message { return newDynamic(desc, t) }
-		return t.add(name, desc, create)
-	}
-	return t.addType(typ)
-}
-
-// addType adds a type by it's reflection type.
-func (t *types) addType(typ reflect.Type) (ty, bool) {
+// addForType adds all types required to serialize the reflection type.
+// The callback will be called for each newly added type.
+func (t *types) addForType(ctx context.Context, typ reflect.Type, cb func(t *ty) error) (*ty, error) {
 	for typ.Kind() == reflect.Ptr {
 		typ = typ.Elem()
 	}
-	create := func() proto.Message { return reflect.New(typ).Interface().(proto.Message) }
-	msg := create()
+	msg := reflect.New(typ).Interface().(proto.Message)
 	name := proto.MessageName(msg)
+	if name == "" {
+		panic(fmt.Errorf("Can not determine the message name of: %v", msg))
+	}
+	if ty, ok := t.byName[name]; ok {
+		return ty, nil
+	}
 	var desc *descriptor.DescriptorProto
 	if d, ok := msg.(protoutil.Described); ok {
 		desc, _ = protoutil.DescriptorOf(d)
 	}
-	return t.add(name, desc, create)
+	ty := t.add(name, desc)
+	if err := cb(ty); err != nil {
+		return nil, err
+	}
+
+	// Recursively add referenced types.
+	numFields := typ.NumField()
```

Comment:
Will panic if the type is not a struct. Why did you need to implement this?
Can we have it tested?

Reply:
We didn't save types properly in the capture, so saving capture would result in file which can not be `gapit unpack`ed.  It could still load gapic as it would just use the run-time descriptors.

So I fixed it so we actually save descriptors properly in go (C++ already does this correctly).

Reply:
Okay. Please check the type is a struct (anything can implement `proto.Message`).
A simple test to cover this would also be good.